### PR TITLE
Add last BuildID to workflow info

### PR DESCRIPTION
**Pull request by Sushisource**
## What was changed
Add the BuildID of the last completed task to info. Importantly this *does not* use the worker's current build id if it isn't in history, as that would be nondeterministic, unlike Binary Checksum which is guaranteed to be set as soon as the first WFT is completed which is not the case here.

## Why?
Useful

## Checklist

1. Part of https://github.com/temporalio/features/issues/253
2. How was this tested:
Added IT

3. Any docs updates needed?

### CODE REVIEW

File: internal/internal_task_handlers.go [hunk 1]
```diff
@@ -267,10 +270,16 @@ func (eh *history) isNextWorkflowTaskFailed() (task finishedTask, err error) {
...
 		var flags []sdkFlag
 		if nextEventType == enumspb.EVENT_TYPE_WORKFLOW_TASK_COMPLETED {
-			binaryChecksum = nextEvent.GetWorkflowTaskCompletedEventAttributes().BinaryChecksum
-			for _, flag := range nextEvent.GetWorkflowTaskCompletedEventAttributes().GetSdkMetadata().GetLangUsedFlags() {
+			completedAttrs := nextEvent.GetWorkflowTaskCompletedEventAttributes()
+			binaryChecksum = completedAttrs.BinaryChecksum
+			workerVersion := completedAttrs.GetWorkerVersion()
+			if workerVersion != nil {
+				buildID = workerVersion.GetBuildId()
+			}
```
Comment (bergundy) [inline internal/internal_task_handlers.go:281]:
```suggestion
			buildID = completedAttrs.GetWorkerVersion().GetBuildId()
```

**Comment by Sushisource** [comment]
Realized I did this a bit of a silly way at first. Draft while I fix it.

File: test/worker_versioning_test.go [hunk 2]
```diff
@@ -273,3 +274,63 @@ func (ts *WorkerVersioningTestSuite) TestReachabilityVersions() {
...
+	ts.NoError(ts.client.SignalWorkflow(ctx, wfHandle.GetID(), wfHandle.GetRunID(), "fakesig", ""))
+	var lastBuildID string
+	enval, err := ts.client.QueryWorkflow(ctx, wfHandle.GetID(), wfHandle.GetRunID(), "get-last-build-id", nil)
+	ts.NoError(err)
+	ts.NoError(enval.Get(&lastBuildID))
+	worker1.Stop()
+	ts.Equal("1.0", lastBuildID)
+
+	// Add new compat ver
+	err = ts.client.UpdateWorkerBuildIdCompatibility(ctx, &client.UpdateWorkerBuildIdCompatibilityOptions{
```
Comment (bergundy) [inline test/worker_versioning_test.go:306]: nit: this doesn't require versioning to test, you can use build ID with unversioned workers.

Comment (bergundy) [reply test/worker_versioning_test.go:306]: Could make sense to test with and without versioning just to be safe.

**Review by bergundy — APPROVED** [review]

File: internal/internal_task_handlers.go (hunk 1 above)
Comment (Sushisource) [reply internal/internal_task_handlers.go]: Already different :)

File: test/worker_versioning_test.go (hunk 2 above)
Comment (Sushisource) [reply test/worker_versioning_test.go:308]: The bit you reviewed is subtly wrong in terms of querying on a cached worker after the worker has completed a task with its' ID. New test also makes sure the first task doesn't have an ID, which covers this path.

File: internal/workflow.go [hunk 3]
```diff
@@ -1016,14 +1020,22 @@ type UpdateInfo struct {
...
 	if wInfo.BinaryChecksum == "" {
 		return getBinaryChecksum()
 	}
 	return wInfo.BinaryChecksum
 }
 
+// GetCurrentBuildID returns the Build ID of the worker that processed this task, which may be
+// empty. During replay this id may not equal the id of the replaying worker. If not replaying and
+// this worker has a defined Build ID, it will equal that ID. It is safe to use for branching.
+func (wInfo *WorkflowInfo) GetCurrentBuildID() string {
```
Comment (Quinn-With-Two-Ns) [inline internal/workflow.go:1035]: May want to add a note for queries

File: internal/internal_task_handlers.go [hunk 4]
```diff
@@ -1016,6 +1026,11 @@ ProcessEvents:
 		} else {
 			w.workflowInfo.BinaryChecksum = binaryChecksum
 		}
+		if isReplay {
+			w.workflowInfo.currentTaskBuildID = currentBuildID
+		} else if !isReplay && !isQueryOnlyTask {
```
Comment (Quinn-With-Two-Ns) [inline internal/internal_task_handlers.go:1031]: can we add a comment here explaining why a query only task is treated special.

**Review by Quinn-With-Two-Ns — APPROVED** [review]

**Pull request merged**